File: jun1/src/combinators/eat_string.py
from ..combinator import Combinator
from ..parse_iteration_result import FrameStack, ParserIterationResult
from ..state import CombinatorState
from ..u8set import U8Set


class EatStringState(CombinatorState):
    def __init__(self, index: int, frame_stack: FrameStack) -> None:
        self.index = index
        self.frame_stack = frame_stack


class EatString(Combinator):
    def __init__(self, string: str) -> None:
        self.string = string

    def initial_state(self, signal_id: int, frame_stack: FrameStack) -> EatStringState:
        return EatStringState(0, frame_stack)

    def next_state(self, state: EatStringState, c: Optional[str], signal_id: int) -> tuple[EatStringState, int, ParserIterationResult]:
        if state.index > len(self.string):
            return state, signal_id, ParserIterationResult(U8Set.none(), False, state.frame_stack)
        if state.index == len(self.string):
            return state, signal_id, ParserIterationResult(U8Set.none(), True, state.frame_stack)
        u8set = U8Set.from_chars(self.string[state.index])
        state.index += 1
        return state, signal_id, ParserIterationResult(u8set, False, state.frame_stack)

File: jun1/src/combinators/eat_string.rs
use crate::combinator::Combinator;
use crate::parse_iteration_result::{FrameStack, ParserIterationResult};
use crate::state::CombinatorState;
use crate::u8set::U8Set;

pub struct EatString(pub &'static str);

impl Combinator for EatString {
    type State = EatStringState;

    fn initial_state(&self, _signal_id: &mut usize, frame_stack: FrameStack) -> Self::State {
        EatStringState {
            index: 0,
            frame_stack,
        }
    }

    fn next_state(&self, state: &mut Self::State, _c: Option<char>, _signal_id: &mut usize) -> ParserIterationResult {
        let state = state.as_any_mut().downcast_mut::<EatStringState>().expect("Invalid state type");
        if state.index > self.0.len() {
            return ParserIterationResult::new(U8Set::none(), false, state.frame_stack.clone());
        }
        if state.index == self.0.len() {
            let mut result =
                ParserIterationResult::new(U8Set::none(), true, state.frame_stack.clone());
            return result;
        }
        let u8set = U8Set::from_chars(&self.0[state.index..=state.index]);
        state.index += 1;
        ParserIterationResult::new(u8set, false, state.frame_stack.clone())
    }
}

pub struct EatStringState {
    pub index: usize,
    pub frame_stack: FrameStack,
}

impl CombinatorState for EatStringState {
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
        self
    }
}
